packages/server/src/ApolloServer.ts

```diff
@@ -807,6 +807,49 @@ export class ApolloServer<in out TContext extends BaseContext = BaseContext> {
         (p) => pluginIsInternal(p) && p.__internal_plugin_id__() === id,
       );
 
+    // Make sure we're not trying to explicitly enable and disable the same
+    // feature. (Be careful: we are not trying to stop people from installing
+    // the same plugin twice if they have a use case for it, like two usage
+    // reporting plugins for different variants.)
+    //
+    // Note that this check doesn't work for the landing page plugin, because
+    // users can write their own landing page plugins and we don't know if a
+    // given plugin is a landing page plugin until the server has started.
+    const pluginsByInternalID = new Map<
+      InternalPluginId,
+      { sawDisabled: boolean; sawNonDisabled: boolean }
+    >();
+    for (const p of plugins) {
+      if (pluginIsInternal(p)) {
+        if (!pluginsByInternalID.has(p.__internal_plugin_id__())) {
+          pluginsByInternalID.set(p.__internal_plugin_id__(), {
+            sawDisabled: false,
+            sawNonDisabled: false,
+          });
+        }
+        if (p.__is_disabled_plugin__()) {
+          pluginsByInternalID.get(p.__internal_plugin_id__())!.sawDisabled =
+            true;
+        } else {
+          pluginsByInternalID.get(p.__internal_plugin_id__())!.sawNonDisabled =
+            true;
+        }
+      }
+    }
+    for (const [
+      id,
+      { sawDisabled, sawNonDisabled },
+    ] of pluginsByInternalID.entries()) {
+      if (sawDisabled && sawNonDisabled) {
+        throw new Error(
+          `You have tried to install both ApolloServerPlugin${id} and ` +
+            `ApolloServerPlugin${id}Disabled in your server. Please choose ` +
+            `whether or not you want to disable the feature and install the ` +
+            `appropriate plugin for your use case.`,
+        );
+      }
+    }
```

Comment:
Pro: this is verbose and it's easy to understand the logic here.

Con: we don't really need this loop, and could even cut the first one short and throw as soon as we see the true/true condition.

Reply:
Sure.